src/lib/tier.ts

```diff
@@ -0,0 +1,220 @@
+import { retryWithDelay } from "@/utils/retry.utils";
+import { aoInstance } from "./aoconnect";
+import { redis } from "./redis";
+
+type TierWallet = {
+  balance: string;
+  rank: number | "";
+  tier: number;
+  progress: number;
+  snapshotTimestamp: number;
+  totalHolders: number;
+};
+
+type WalletsTierInfo = Record<string, TierWallet>;
+
+type CachedWalletsTierInfo = {
+  walletsTierInfo: WalletsTierInfo;
+  snapshotTimestamp: number;
+  totalWallets: number;
+};
+
+const TierTypes = {
+  PRIME: 1,
+  EDGE: 2,
+  RESERVE: 3,
+  SELECT: 4,
+  CORE: 5,
+} as const;
+
+const TierNames = {
+  1: "Prime",
+  2: "Edge",
+  3: "Reserve",
+  4: "Select",
+  5: "Core",
+} as const;
+
+const Tiers = [
+  // Prime tier (Top 2%)
+  {
+    name: TierNames[TierTypes.PRIME],
+    thresholdPercent: 2, // Top 2% of wallets
+    progressMin: 98, // Progress range: 98-100%
+    progressMax: 100,
+  },
+  // Edge tier (Top 2.01% to 20%)
+  {
+    name: TierNames[TierTypes.EDGE],
+    thresholdPercent: 20, // Top 20% of wallets
+    progressMin: 80, // Progress range: 80-98%
+    progressMax: 98,
+  },
+  // Reserve tier (Top 20.01% to 50%)
+  {
+    name: TierNames[TierTypes.RESERVE],
+    thresholdPercent: 50, // Top 50% of wallets
+    progressMin: 50, // Progress range: 50-80%
+    progressMax: 80,
+  },
+  // Select tier (Top 50.01% to 80%)
+  {
+    name: TierNames[TierTypes.SELECT],
+    thresholdPercent: 80, // Top 80% of wallets
+    progressMin: 20, // Progress range: 20-50%
+    progressMax: 50,
+  },
+  // Core tier (Below 80.01%)
+  {
+    name: TierNames[TierTypes.CORE],
+    thresholdPercent: 100, // Everyone (bottom 20%)
+    progressMin: 0, // Progress range: 0-20%
+    progressMax: 20,
+  },
+];
+
+function getTierThresholds(totalWallets: number) {
+  const tierThresholds = [];
+
+  if (totalWallets > 0) {
+    for (let i = 0; i < Tiers.length; i++) {
+      const tier = Tiers[i];
+      const tierIndex = i + 1; // Convert to 1-based index for tier comparison
+      const tierThreshold = {
+        maxRank: Math.ceil((tier.thresholdPercent * totalWallets) / 100),
+        minRank:
+          tierIndex == TierTypes.PRIME
+            ? 1
+            : tierIndex > TierTypes.PRIME
+            ? Math.ceil((Tiers[i - 1].thresholdPercent * totalWallets) / 100) +
+              1
+            : 1,
+      };
+      tierThresholds.push(tierThreshold);
+    }
+  }
+
+  return tierThresholds;
+}
+
+/**
+ * Calculate progress percent within tier range
+ * @param walletRank The rank of the wallet
+ * @param totalWallets Total number of wallets
+ * @returns Progress percentage within tier range
+ */
+function calculateTierProgressPercent(
+  walletRank: number,
+  totalWallets: number
+): number {
+  if (walletRank <= 0 || totalWallets <= 0) return 0;
+  return (
+    Math.floor(
+      ((totalWallets - walletRank + 1) / totalWallets) * 100 * Math.pow(10, 6)
+    ) / Math.pow(10, 6)
```

Comment:
Why multiply and divide by the same number?

Reply:
To get the number to 6 decimal places.

Reply:
Sorry! You are right! I didn't see that you were flooring the multiplication and then dividing 🤦